File: tests/program-workflow.test.py
import os,sys,pathlib,tempfile,subprocess,base64,struct,zlib,json

with tempfile.TemporaryDirectory(prefix='pve-program-') as tmp:
 os.environ['PVE_DATA']=tmp;root=pathlib.Path(__file__).resolve().parent.parent
 sys.path.insert(0,str(root/'engine'));import core
 source=pathlib.Path(tmp)/'source.mp4'
 subprocess.run(core.BASE+['-f','lavfi','-i','color=c=red:s=640x360:r=30:d=6','-f','lavfi','-i','sine=frequency=440:duration=6','-c:v','libx264','-pix_fmt','yuv420p','-c:a','aac','-shortest',source],check=True,capture_output=True)
 media=core.inspect(source,'program-source',source.name)
 def png(color,width=240,height=80,align='center'):
  rows=b''.join(b'\0'+bytes(color)*width for _ in range(height))
  def chunk(kind,data):return struct.pack('!I',len(data))+kind+data+struct.pack('!I',zlib.crc32(kind+data)&0xffffffff)
  payload=b'\x89PNG\r\n\x1a\n'+chunk(b'IHDR',struct.pack('!2I5B',width,height,8,6,0,0,0))+chunk(b'IDAT',zlib.compress(rows))+chunk(b'IEND',b'')
  anchor={'left':8,'right':width-8}.get(align,width/2)
  return {'data':'data:image/png;base64,'+base64.b64encode(payload).decode(),'width':width,'height':height,'anchorX':anchor,'anchorY':height/2}
 clip={'id':'c','media':media['id'],'in':0,'out':6,'start':0,'layer':0,'speed':1,'endSpeed':1,'curve':'constant','audio':{'volume':1},'color':{},'stabilization':'OFF'}
 texts=[
  {'id':'title','text':'TITLE','start':0,'end':2.2,'size':88,'x':.5,'y':.5,'align':'center','opacity':1,'raster':png((255,255,255,255))},
  {'id':'line-a','text':'LINE A','start':2.5,'end':3.3,'size':68,'x':.5,'y':.82,'align':'center','opacity':1,'raster':png((255,230,40,255))},
  {'id':'line-b','text':'LINE B','start':3.3,'end':4.2,'size':68,'x':.08,'y':.82,'align':'left','opacity':1,'raster':png((70,210,255,255),align='left')}
 ]
 project={'version':1,'media':[media],'clips':[clip],'audioClips':[],'videoTracks':[{'hidden':False} for _ in range(3)],'audioTracks':[{'volume':1} for _ in range(4)],'bgm':{},'texts':texts,'effects':[{'type':'black-in','start':0,'duration':2.5,'hold':1.5,'strength':1},{'type':'black-out','start':4.5,'duration':1.5,'hold':.3,'strength':1}],'aspect':'16:9','export':{'preset':'YOUTUBE','fps':'30','resolution':'Source','quality':'High','codec':'H.264'}}
 result=core.render({'id':'program','cancel':False},project);out=core.ROOT/'exports'/result['file'];core.validate_output(out,6,True)
 def pixel(t,x,y):return list(subprocess.check_output(core.BASE+['-v','error','-ss',str(t),'-i',out,'-frames:v','1','-vf',f'crop=4:4:{x}:{y},scale=1:1','-pix_fmt','rgb24','-f','rawvideo','pipe:1']))
 corner=[pixel(t,10,10) for t in (.5,1.75,2.25,2.6,5.1,5.8)]
 assert max(corner[0])<20,corner
 assert 35<corner[1][0]<100 and corner[1][1]<20,corner
 assert 150<corner[2][0]<225 and corner[2][1]<30,corner
 assert corner[3][0]>220 and corner[3][1]<30,corner
 assert 75<corner[4][0]<180 and corner[4][1]<30,corner
 assert max(corner[5])<25,corner
 assert min(pixel(.5,320,180))>225
 yellow=pixel(2.8,320,295);assert yellow[0]>200 and yellow[1]>170 and yellow[2]<100,yellow
 blue=pixel(3.6,55,295);assert blue[2]>180 and blue[1]>150 and blue[0]<130,blue
 # At the exact caption boundary, the first line is gone and the second line is active.
 boundary=pixel(3.3,55,295);assert boundary[2]>150,boundary
 probe=json.loads(subprocess.check_output([core.FFPROBE,'-v','error','-show_streams','-of','json',out]));video=next(s for s in probe['streams'] if s['codec_type']=='video')
 assert (video['width'],video['height'],video['codec_name'])==(640,360,'h264')
 print('YouTube program export: held black title, exact caption swaps, picture fade-in/out, size/position and full decode PASS')

File: tests/program-workflow.test.mjs
import assert from 'node:assert/strict';
import {project,clip,sanitize} from '../dist/model.js';
import {effectAlpha} from '../dist/creative.js';
import {sequentialCaptions,youtubeProgramPlan,applyYoutubeProgram} from '../dist/program-workflow.js';
import {captionPresets} from '../dist/caption-presets.js';
import {clampTimelineHeight,timelineRatio} from '../dist/workspace-splitter.js';

const ids=['a','b','c'];
const captions=sequentialCaptions('最初のセリフ\n\n次のセリフ\n強調',{start:2,duration:1.5,gap:.1,limit:8,id:()=>ids.shift(),style:captionPresets.varietyYellow.style});
assert.deepEqual(captions.map(value=>[value.id,value.text,value.start,value.end]),[['a','最初のセリフ',2,3.5],['b','次のセリフ',3.6,5.1],['c','強調',5.199999999999999,6.699999999999999]]);
assert.equal(captions[0].outline,7);assert.equal(captions[0].color,'#ffe64d');

const plan=youtubeProgramPlan(10,{title:'番組タイトル',hold:1.5,fade:1,outro:2});
const intro=plan.effects[0],outro=plan.effects[1];
assert.equal(effectAlpha(intro,.5),1);assert.equal(effectAlpha(intro,1.5),1);assert(Math.abs(effectAlpha(intro,2)-.5)<1e-9);assert.equal(effectAlpha(intro,2.5),0);
assert.equal(effectAlpha(outro,8),0);assert(effectAlpha(outro,9)>.5);assert.equal(effectAlpha(outro,9.9),1);
assert.equal(plan.text.text,'番組タイトル');assert.equal(plan.aspect,'16:9');assert.equal(plan.export.preset,'YOUTUBE');

const p=project(),media={id:'v',name:'video',kind:'video',duration:10};p.media=[media];p.clips=[clip(media)];const applied=applyYoutubeProgram(p,{title:'TEST'});
assert.equal(p.aspect,'16:9');assert.equal(p.effects.length,2);assert.equal(p.texts.length,1);assert.equal(applied.text.text,'TEST');
const restored=sanitize(JSON.parse(JSON.stringify(p)));assert.equal(restored.effects[0].hold,1.6);assert(restored.effects[1].hold>0);

assert.equal(clampTimelineHeight(50,800),190);assert.equal(clampTimelineHeight(900,800),610);assert(Math.abs(timelineRatio(500,100,900)-.5)<1e-9);
console.log('YouTube program flow: black hold/fades, sequential variety captions and workspace splitter PASS');
